Data/ByteString/Builder/RealFloat/Internal.hs

```diff
@@ -0,0 +1,438 @@
+  | v >= 1000000000 = 10
+  | v >= 100000000 = 9
+  | v >= 10000000 = 8
+  | v >= 1000000 = 7
+  | v >= 100000 = 6
+  | v >= 10000 = 5
+  | v >= 1000 = 4
+  | v >= 100 = 3
+  | v >= 10 = 2
+  | otherwise = 1
+
+infixr 9 `dot`
+dot :: (a -> b) -> (c -> d -> a) -> c -> d -> b
+dot = (.) . (.)
+
+infixr 9 ...
+(...) :: (a -> b) -> (c -> d -> e -> a) -> c -> d -> e -> b
+(...) = dot . (.)
+
+-- TODO encode proof in types?
+-- From 'In-and-Out Conversions' https://dl.acm.org/citation.cfm?id=362887, we
+-- have that a conversion from a base-b n-digit number to a base-v m-digit
+-- number such that the round-trip conversion is identity requires
+--
+--    v^(m-1) > b^n
+--
+-- Specifically for binary floating point to decimal conversion, we must have
+--
+--    10^(m-1) > 2^n
+-- => log(10^(m-1)) > log(2^n)
+-- => (m-1) * log(10) > n * log(2)
+-- => m-1 > n * log(2) / log(10)
+-- => m-1 >= ceil(n * log(2) / log(10))
+-- => m >= ceil(n * log(2) / log(10)) + 1
+--
+-- And since 32 and 64-bit floats have 23 and 52 bits of mantissa (and then an
+-- implicit leading-bit), we need
+--
+--    ceil(24 * log(2) / log(10)) + 1 => 9
+--    ceil(53 * log(2) / log(10)) + 1 => 17
+--
+-- In addition, the exponent range from floats is [-45,38] and doubles is
+-- [-324,308] (including subnormals) which are 3 and 4 digits respectively
+--
+-- Thus we have,
+--
+--    floats: 1 (sign) + 9 (mantissa) + 1 (.) + 1 (e) + 3 (exponent) = 15
+--    doubles: 1 (sign) + 17 (mantissa) + 1 (.) + 1 (e) + 4 (exponent) = 24
+--
+maxEncodedLength :: Int
+maxEncodedLength = 32
+
+-- TODO TH?
+pokeAll :: String -> Ptr Word8 -> IO (Ptr Word8)
+pokeAll s ptr = foldM pokeOne ptr s
+  where pokeOne p c = S.poke p (c2w c) >> return (p `plusPtr` 1)
+
+boundString :: String -> BoundedPrim ()
+boundString s = boundedPrim maxEncodedLength $ const (pokeAll s)
+
+--         Sign -> Exp  -> Mantissa
+special :: Bool -> Bool -> Bool -> BoundedPrim ()
+special    _       _       True =  boundString "NaN"
+special    True    False   _    =  boundString "-0.0e0"
+special    False   False   _    =  boundString "0.0e0"
+special    True    True    _    =  boundString "-Infinity"
+special    False   True    _    =  boundString "Infinity"
+
+-- Returns e == 0 ? 1 : ceil(log_2(5^e)); requires 0 <= e <= 3528.
+pow5bitsUnboxed :: Int# -> Int#
+pow5bitsUnboxed e = (e *# 1217359#) `uncheckedIShiftRL#` 19# +# 1#
+
+-- Returns floor(log_10(2^e)); requires 0 <= e <= 1650.
+log10pow2Unboxed :: Int# -> Int#
+log10pow2Unboxed e = (e *# 78913#) `uncheckedIShiftRL#` 18#
+
+-- Returns floor(log_10(5^e)); requires 0 <= e <= 2620.
+log10pow5Unboxed :: Int# -> Int#
+log10pow5Unboxed e = (e *# 732928#) `uncheckedIShiftRL#` 20#
+
+acceptBoundsUnboxed :: Word# -> Int#
+acceptBoundsUnboxed _ = 0#
+-- for round-to-even and correct shortest
+-- acceptBoundsUnboxed v = ((v `uncheckedShiftRL#` 2#) `and#` 1##) `eqWord#` 0##
+
+fquot10 :: Word# -> Word#
+fquot10 w = (w `timesWord#` 0xCCCCCCCD##) `uncheckedShiftRL#` 35#
```

Comment:
Magic constants for division here and below require either an explanation why it works, or a reference to some source. For instance, I'd point readers to `https://hackage.haskell.org/package/quote-quot-0.2.0.0/docs/Numeric-QuoteQuot.html#v:astQuot` as a mean of verification.